### ISSUE TITLE
Sphinx 3.1.0 import error

### ISSUE THREAD

**Issue by larsoner** [sphinx-doc/sphinx#7810]
Over in SciPy we're getting build failures [of the form](https://app.circleci.com/pipelines/github/scipy/scipy/1842/workflows/74345bb6-2032-46e0-ac87-562e56cb8bc7/jobs/19769/steps):
```
LANG=C "/home/circleci/repo/venv/bin/python" -msphinx -WT --keep-going -t scipyorg  -b html -d build/doctrees  "-j2" source build/html-scipyorg 
Running Sphinx v3.1.0

Traceback (most recent call last):
  File "/home/circleci/repo/venv/lib/python3.7/site-packages/sphinx/config.py", line 319, in eval_config_file
    execfile_(filename, namespace)
  File "/home/circleci/repo/venv/lib/python3.7/site-packages/sphinx/util/pycompat.py", line 88, in execfile_
    exec(code, _globals)
  File "/home/circleci/repo/doc/source/conf.py", line 21, in <module>
    from sphinx.util import inspect
  File "/home/circleci/repo/venv/lib/python3.7/site-packages/sphinx/util/inspect.py", line 28, in <module>
    from sphinx.pycode.ast import ast  # for py35-37
  File "/home/circleci/repo/venv/lib/python3.7/site-packages/sphinx/pycode/__init__.py", line 23, in <module>
    from sphinx.pycode.parser import Parser
  File "/home/circleci/repo/venv/lib/python3.7/site-packages/sphinx/pycode/parser.py", line 22, in <module>
    from sphinx.util.inspect import signature_from_ast
ImportError: cannot import name 'signature_from_ast' from 'sphinx.util.inspect' (/home/circleci/repo/venv/lib/python3.7/site-packages/sphinx/util/inspect.py)
```
Could this be a Sphinx 3.1.0 packaging problem? Nothing changed AFAIK at the SciPy end...

**Comment by larsoner** [comment]
... actually this looks like a circular import problem

**Comment by larsoner** [comment]
I can confirm this on the `3.1.x` branch with just:
```
larsoner@bunk:~/python/sphinx$ python -c "from sphinx.util import inspect"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/larsoner/python/sphinx/sphinx/util/inspect.py", line 28, in <module>
    from sphinx.pycode.ast import ast  # for py35-37
  File "/home/larsoner/python/sphinx/sphinx/pycode/__init__.py", line 23, in <module>
    from sphinx.pycode.parser import Parser
  File "/home/larsoner/python/sphinx/sphinx/pycode/parser.py", line 22, in <module>
    from sphinx.util.inspect import signature_from_ast
ImportError: cannot import name 'signature_from_ast' from partially initialized module 'sphinx.util.inspect' (most likely due to a circular import) (/home/larsoner/python/sphinx/sphinx/util/inspect.py)
```

**Comment by tk0miya** [comment]
Fixed by #7811 
Thank you for reporting & contributing!
